includes/class-wc-order-product-lookup-helper.php

```diff
@@ -0,0 +1,208 @@
+ *
+ * @author   Automattic
+ * @package  WooCommerce
+ * @since    3.3.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	// Exit if accessed directly.
+	exit;
+}
+
+/**
+ * WC_Order_Product_Lookup_Helper class.
+ *
+ * @class     WC_Order_Product_Lookup_Helper
+ * @version   3.3.0
+ * @package   WooCommerce/Classes
+ * @category  Class
+ * @author    WooThemes
+ */
+class WC_Order_Product_Lookup_Helper {
+
+	/**
+	 * Hook in methods.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function init() {
+		add_action( 'woocommerce_delete_order', array( __CLASS__, 'handle_delete_order' ), 10, 1 );
+		add_action( 'woocommerce_delete_product', array( __CLASS__, 'handle_delete_product' ), 10, 1 );
+		add_action( 'woocommerce_delete_product_variation', array( __CLASS__, 'handle_delete_product' ), 10, 1 );
+		add_action( 'woocommerce_after_order_object_save', array( __CLASS__, 'handle_new_order' ), 10, 1 );
+		add_action( 'woocommerce_update_order', array( __CLASS__, 'handle_update_order' ), 10, 1 );
+	}
+
+	/**
+	 * Handler for deleting an order, keep the lookup table in sync.
+	 *
+	 * @param int $order_id Order ID.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_delete_order( $order_id ) {
+		global $wpdb;
+
+		$wpdb->delete(
+			$wpdb->prefix . 'woocommerce_order_product_lookup',
+			array(
+				'order_id' => $order_id,
+			),
+			array(
+				'%d',
+			)
+		); // WPCS: cache ok.
+	}
+
+	/**
+	 * Handler for deleting a product, keep the lookup table in sync.
+	 *
+	 * @param int $product_id Product ID.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_delete_product( $product_id ) {
+		global $wpdb;
+
+		$wpdb->delete(
+			$wpdb->prefix . 'woocommerce_order_product_lookup',
+			array(
+				'product_id' => $product_id,
+			),
+			array(
+				'%d',
+			)
+		); // WPCS: cache ok.
+	}
+
+	/**
+	 * Handler for saving an order, keep the lookup table in sync.
+	 *
+	 * @param WC_Order $order Order object.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_new_order( $order ) {
+		global $wpdb;
+
+		$order_items = $order->get_items();
+
+		foreach ( $order_items as $item_id => $item ) {
+			$item_product_id = $item->get_product_id();
+
+			// Skip if this line item doesn't have a product id associated.
+			if ( empty( $item_product_id ) ) {
+				continue;
+			}
+
+			$data = array(
+				'user_email'           => $order->get_billing_email(),
+				'order_id'             => $order->get_id(),
+				'user_id'              => $order->get_customer_id(),
+				'product_id'           => $item_product_id,
+				'parent_order_item_id' => $item_id,
+			);
+
+			$wpdb->insert(
+				$wpdb->prefix . 'woocommerce_order_product_lookup',
+				$data,
+				array(
+					'%s',
+					'%d',
+					'%d',
+					'%d',
+					'%d',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Handler for updating an order, keep the lookup table in sync.
+	 *
+	 * @param int $order_id Order ID.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_update_order( $order_id ) {
```

Comment:
If an order gets saved multiple times, this could result in a lot of delete/insert events in close proximity. Can we defer this sync somehow so it runs less often/is non-blocking?